torch_xla/experimental/xla_sharding.py

```diff
@@ -521,3 +521,42 @@ def apply(self, t: torch.Tensor):
     # TODO(yeounoh) use virtual device interface when available.
     assert (t.device == xm.xla_device())
     mark_sharding(t, self.mesh, self.partition_spec)
+
+
+class XLAPatchedLinear(torch.autograd.Function):
+  """
+  A patched version of `torch.nn.functional.linear` that uses einsum instead
+  of torch.matmul which will flatten the tensors to 2D and collide the sharded
+  dimensions. The torch.matmul default behavior makes it very hard for XLA compiler
+  to propagate the sharding annotation.
+
+  TODO (alanwaketan): Let's patch it on the dispatcher level.
+  """
+
+  @staticmethod
+  def forward(ctx, input, weight, bias=None):
+    # bias is an optional argument
+    ctx.save_for_backward(input, weight, bias)
```

Comment:
hmm.. are we doing explicit gradient checkpointing here?

Reply:
not family with how nn moudle is being written but I mainly ask because the `torch.no_grad` below.

Reply:
No, it's just a convention in autograd function to save the activations. So not the gradient checkpointing or re-materialization.